File: nodejs/lesson04/solution/formatter.js
'use strict';

const express = require('express')
const app = express()
const initTracer = require('../../lib/tracing').initTracer;
const { Tags, FORMAT_HTTP_HEADERS } = require('opentracing');

const tracer = initTracer('format-service');

const port = 8081;

app.listen(port, function () {
    console.log('Formatter app listening on port ' + port);
})

app.get('/format/:str', function (req, res) {
    const parentSpanContext = tracer.extract(FORMAT_HTTP_HEADERS, req.headers)
    const span = tracer.startSpan('http_server', {
        childOf: parentSpanContext
    });
    span.setTag(Tags.SPAN_KIND, Tags.SPAN_KIND_RPC_SERVER);

    const str= req.params.str;

    span.log({
        'event': 'format',
        'value': str
    });

    //this requires downstream service to know about the baggage content
    let greeting = span.getBaggageItem('greeting') || 'Hello';

    span.finish();

    res.send(`${greeting}, ${str}`);
})

File: python/lesson04/solution/formatter.py
from flask import Flask
from flask import request
from lib.tracing import init_tracer
from opentracing.ext import tags
from opentracing.propagation import Format

app = Flask(__name__)
tracer = init_tracer('formatter') 
 
@app.route("/format")
def format():
    span_ctx = tracer.extract(Format.HTTP_HEADERS, request.headers)
    span_tags = {tags.SPAN_KIND: tags.SPAN_KIND_RPC_SERVER}
    with tracer.start_active_span('format', child_of=span_ctx, tags=span_tags) as scope:
        greeting = scope.span.get_baggage_item('greeting')
        if not greeting:
            greeting = 'Hello'
        hello_to = request.args.get('helloTo')
        return '%s, %s!' % (greeting, hello_to)

if __name__ == "__main__":
    app.run(port=8081)
